### ISSUE TITLE
Celebrity Voice Matching

### ISSUE THREAD

**Issue by msj121** [speechbrain/speechbrain#1359]
Is there a way to match an encoding to the voxelceleb database. Analyze a voice and know which celebrity it is? Thanks

All I could see was comparing two voices with encodings.

Currently:
```

import speechbrain as sb

import torchaudio
from speechbrain.pretrained import EncoderClassifier


classifier = EncoderClassifier.from_hparams(source="speechbrain/spkrec-ecapa-voxceleb")
signal, fs =torchaudio.load('example1.wav')
embeddings = classifier.encode_batch(signal)

print(embeddings)
```

Will print: 

```
tensor([[[-21.4220,  -7.9435,  -9.8436, -26.7140,   9.0414,   2.7003,  16.9070,
            2.7529, -18.0246,  -8.3216, -15.7110,  29.2359,  22.8233,  -4.9595,
           -7.4860, -33.5905,   8.2758,  17.1753,  30.2007, -26.1966,  17.6238,
            0.7029,  10.6952,  15.7721, -11.1679,  18.9203,  12.1647,  -1.1949,
           -0.6563, -54.0647,  14.0799,  -4.5127,   1.1240,  10.5670,   1.7021,
           -1.5989,  19.8689,  -2.2041, -11.4251,  12.6044, -10.7529,   2.6609,
           22.3877, -10.8091,  25.6252, -20.2354,   2.8001,  20.6080,  -2.4577,
          -29.3553, -10.6289,  -5.4961, -18.3435,  -4.3220,  24.6399,  10.4924,
           -7.7208, -14.2377,  -8.7558,  46.1296,  -8.7363,  -4.1992,  13.2267,
           13.3115,  51.5118, -18.2723,  14.6351, -49.1952,  -6.1234,   4.1286,
          -13.4014, -24.9110,  34.3119, -47.2819, -48.0277,  18.8033,  25.1597,
          -12.7326,   5.7272,  15.9076,  14.0736,   3.0563,  -3.2419,  35.0953,
            9.7102,  38.8568, -18.3337,  21.4572, -13.9801,  -2.7211,  -8.1815,
            4.9559,   9.1505,  24.2812, -21.6174,   8.7041,  11.4554,  -8.4128,
            4.2390,  -8.6488,  23.9885, -37.8855,  -5.3550,  22.7303, -21.3078,
           23.6596, -30.8486, -26.8964, -15.6380,  -9.4559,  34.3659,  -0.1649,
          -53.8423,  19.9297, -37.3943,  14.9680,  19.2769,   8.0847,  21.8415,
          -13.1034, -16.0090,  36.6997,  21.7576,  20.1418,  56.6040,   6.1807,
           22.4875,  -1.9395, -17.7474,   8.4440, -25.2567, -21.7312,  22.8819,
           31.2812, -37.2362, -51.4757,  53.3587,  34.4824, -23.4280, -18.7449,
           31.1060, -56.6037, -22.1150,  12.3340, -12.6651,  -9.7937,  -9.4300,
            8.1099,  19.2551, -54.0380,  12.1325,   4.0342, -13.3209,  -8.9325,
          -13.0977,  10.1265,  13.5028,   5.3720,  14.7282, -18.1057,  18.2131,
           21.7648,   0.2792,   7.3133, -26.2929,  16.3130,  11.7941,   6.1783,
          -30.1440, -35.8928,   1.3776, -25.7713,  -8.4900,  -2.6336,  16.9865,
           25.7273,  -5.7889,   7.6070,   3.1014,  17.1161,  39.6544, -57.7410,
           23.5208, -11.5448,   3.3124,  34.3362,   3.2499,   7.1039,  -5.1438,
           17.3884,   5.7022,  -1.9973]]])
```

I know I can compare distances from an encoding to another, but is that the only option? Is there a quicker way to determine which of the 5000 celebrities it is? Thanks for any help

**Comment by anautsch** [comment]
`encode_batch` returns an audio representation but not a classification score of dis/similarity. The embedding printed above are a fixed-length audio representation (from duration-variable audio) by a voice biometric model which is provided via [HuggingFace](https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb) with audio normalisation; TDNN embedding extraction, and TDNN classifier.

There is a difference between biometric tasks: verification (1:1 comparison) and identification (1:n comparison). One can also extend to clustering issues (m:n concerns). The [VoxCeleb recipe](https://github.com/speechbrain/speechbrain/tree/develop/recipes/VoxCeleb/SpeakerRec) provides cosine & PLDA socring. Both rely on the TDNN embedding extraction and require to create an enrolment database of reference embeddings per speaker (keep in mind, biometric profiling relates to privacy and consent). Each scoring has their way of handling reference embeddings. TDNN+cosine treats references implicitly by hard-wiring them into the TDNN classifier - it's the referenced example.

In `speechbrain.pretrained.interfaces`, the `EncoderClassifier` and `SpeakerRecognition` classes are documented with code snippets, e.g., `prediction =  classifier.classify_batch(signal)` provides four values for the `example1.wav` data which you might be interested in. The `SpeakerRecognition` class provides functions for 1:n and m:n comparisons.

Are you working on the provided VoxCeleb recipe or on another (public) database?
Please open a PR to improve SpeechBrain :)

**Comment by msj121** [comment]
Yes I am working with recipe trying to explore speaker identification with recipe. Will try classify_batch. Is there a way I can compute "distance" if I make my own database of encodings? Thanks.

**Comment by anautsch** [comment]
The notion of 'distance' depends on the type of score returned. PLDA scores likely encode the trade-off between similarity and dissimilarity at once (with larger value = more similar) in so-called LLRs, whereas other scores might be purely concerned about distance or closeness.

If you do not intend to retrain but make use of the pretrained models, there are certainly different ways to try out - this is my personal perspective only:
* use PLDA and adjust enrol/test protocols (recommended but likely worse performance on standard datasets)
* use the ECAPA-TDNN classifier; ignore the softmax and take the layer before -> train your own classifier based on these "embeddings"

How are you planning to organise your database to check if the recipe did well?
(your train/dev/validation protocols for enrol/test data)

**Comment by msj121** [comment]
Perfect thanks!
